src/VectorArrows3d.ts

```diff
@@ -0,0 +1,349 @@
+    newWorldScale = this.meshPivot.getWorldScale(newWorldScale);
+
+    // Scale is inverted on mesh pivot to cancel out parent transforms (though
+    // translation and rotation are still affected by any parent transforms).
+    // This allows arrows meshes to be scaled 1:1 with world space, regardless
+    // of parent transforms, and prevents distortion or skewing of the mesh.
+    // Parent scaling is  applied to arrow positions and deltas (see
+    // `updateAllArrowTransforms`), rather than the meshes themselves.
+    const invertScale = new Vector3(1, 1, 1).divide(newWorldScale);
+    this.meshPivot.scale.copy(invertScale);
+
+    if (!newWorldScale.equals(this.worldScale)) {
+      this.worldScale.copy(newWorldScale);
+      if (this.positions && this.deltas) {
+        this.setArrowData(this.positions, this.deltas);
+      }
+    }
+  }
+
+  private updateSingleArrowTransform(index: number, src: Vector3, delta: Vector3, diameter: number): void {
+    // Update the arrow shaft
+    const headHeight = HEAD_BASE_HEIGHT * diameter;
+    const length = delta.length();
+    const shaftHeight = Math.max(length - headHeight, 0);
+    if (shaftHeight < 1e-6) {
+      // If the shaft height is too small, scale to 0.
+      this.tempScale.set(0, 0, 0);
+    } else {
+      this.tempScale.set(diameter, diameter, shaftHeight);
+    }
+
+    this.tempMatrix.scale.copy(this.tempScale);
+    this.tempMatrix.position.copy(src);
+    this.tempDst.copy(src).add(delta);
+    this.tempMatrix.lookAt(this.tempDst);
+    this.tempMatrix.updateMatrix();
+    this.shaftInstancedMesh.setMatrixAt(index, this.tempMatrix.matrix);
+
+    if (length < headHeight) {
+      // If head is longer than the total length, shrink the head to match
+      // length. TODO: Is it okay to do this automatically?
+      const newDiameter = length / HEAD_BASE_HEIGHT;
+      this.tempScale.set(newDiameter, newDiameter, newDiameter);
+    } else {
+      this.tempScale.set(diameter, diameter, diameter);
+    }
+    this.tempMatrix.scale.copy(this.tempScale);
+    this.tempMatrix.position.copy(this.tempDst);
+    this.tempDst.add(delta);
+    this.tempMatrix.lookAt(this.tempDst);
+    this.tempMatrix.updateMatrix();
+    this.headInstancedMesh.setMatrixAt(index, this.tempMatrix.matrix);
+  }
+
+  private updateAllArrowTransforms(): void {
+    if (!this.positions || !this.deltas) {
+      return;
+    }
+    const count = this.positions.length / 3;
+    const combinedScale = new Vector3().copy(this.scale).multiply(this.flipAxes).multiply(this.worldScale);
+
+    const tempSrc = new Vector3();
+    const tempDelta = new Vector3();
+    let tempDiameter: number;
+    for (let i = 0; i < count; i++) {
+      // Points and deltas scaled to volume space.
+      tempSrc.fromArray(this.positions, i * 3).multiply(combinedScale);
+      tempDelta.fromArray(this.deltas, i * 3).multiply(combinedScale);
+      // Fallback in case 0-length diameter array is provided
```

Comment:
is a 0-length diameter array provided often?  if so, maybe the comment could say why

Reply:
It shouldn't, but it's possible? 

One scenario I can see happening is if there are no vector arrows on the current frame, so the positions, deltas, and diameter arrays are all empty. And then a later API call only provides positions and deltas.

I can also just remove the comment!